openerp/fields.py

```diff
@@ -861,8 +861,16 @@ def __set__(self, record, value):
             env.invalidate(spec)
 
         else:
-            # simply write to the database, and update cache
+            # Write to database
             record.write({self.name: self.convert_to_write(value)})
+            try:
+                # If there were any NewId records, invalidate cache
+                if len(value.ids) < len(value._ids):
+                    record.modified([self.name])
+                    return
+            except AttributeError:
+                pass
+            # Update cache with new value
```

Comment:
The method `write()` already calls `modified` on every modified field.
If you want to avoid updating the cache when `value` contains new records, do:
```python
# update the cache unless value contains a new record
if not(isinstance(value, BaseModel) and not all(value._ids)):
    record._cache[self] = value
```